### ISSUE TITLE
Allow editing the authentication token only in the config file

### ISSUE THREAD

**Issue by blochl** [rclone/rclone#2627]
#### What is your current rclone version (output from `rclone version`)?

`1.43.1`

#### What problem are you are trying to solve?

Being able to modify the authentication token __only__ in the config file.

##### Use case:

The same configurations' file needs to be distributed between several people who have a shared cloud drive, so that their configurations would be identical. The config file can not be shared with the token included for security reasons, and also because each person should authenticate as themselves, and not to use some common token.

Currently, ground-up configuration for every member is necessary.

#### How do you think rclone should be changed to solve that?

Make it possible to edit the access token only (or re-authenticate). The command `rclone authorize` may be used, but it needs to display the link for the authentication, and not merely try to open the browser, because the server on which it runs may be headless.

**Comment by ncw** [comment]
Let's say you distribute the config file without the token.

Your users can run `rclone config` and edit the remote answering `n` to whether you wish to edit each value.

After that rclone will ask you whether you want to refresh the token.

What do you think of that work-flow?

**Comment by blochl** [comment]
@ncw , Thanks for your suggestion! It slipped my mind that Rclone asks to modify the token **after** the edit is done.

However, the workflow you suggested is not ideal: it can not be automized, unless tools like Expect are used, but that would be a fragile solution, since the questions can vary in each version. Also, after the question whether I would like to change the token, it asks whether I would like to change the team drive, and this is not needed.

Is there a way to just generate the token, without any interaction besides entering the resulting key from Google API after copying the displayed link to the browser? (`rclone authorize` style, just without relying that there is a browser installed on the server)

**Comment by ncw** [comment]
Hmm, perhaps this workflow would work

Use `rclone authorize` to generate a token, then use `rclone config update remoteName token "the token"`

**Comment by blochl** [comment]
Thanks for the tip again - I wasn't aware of `rclone config update remoteName token "the token"`.

And the last workflow you suggest would be perfect, if `rclone authorize` would show the link to generate the authentication code, and wait for the code input. Currently, `rclone authorize` is acting as configuration in `auto config` mode, and assumes that there is a browser installed on the machine where it is run.

If `rclone authorize` had an option to just show the activation link, wait for the auth code, and generate the token, without even trying to open the browser or activate a local server, that would be perfect.

**Comment by ncw** [comment]
Can the users not run `rclone authorize` on their local machines - that is what it was designed for?  Then past the token into that command on the remote machine and they are good to go.

**Comment by blochl** [comment]
@ncw , thanks again, but requesting the users to run `rclone authorize` on their local machines is not an option - there is no guarantee that Rclone is installed there, and there can not be a requirement to install it.

I realize that `rclone authorize` is designed for the scenario that you've described, but is it possible to add a non-auto option to it, _e.g._ `rclone authorize noauto`? This will enable automatic configuration on a headless server with the only manual interaction of going to a link and providing the code, without switching terminals, and without Rclone installed locally. Currently such a functionality can only be achieved with `expect`, which is really not robust.

**Comment by ncw** [comment]
> I realize that `rclone authorize` is designed for the scenario that you've described, but is it possible to add a non-auto option to it, _e.g._ `rclone authorize noauto`? This will enable automatic configuration on a headless server with the only manual interaction of going to a link and providing the code, without switching terminals, and without Rclone installed locally. Currently such a functionality can only be achieved with `expect`, which is really not robust.

If you are using google drive only, then rclone could print a URL for the user to go to to get a code to paste back into rclone - is that what you are thinking of?  The auth for drive used to work like that anyway so it shouldn't be too hard to make it work like that again (optionally).

That isn't a solution for any other provider though, because as far as I know they make you go through the full redirect flow for oauth.

**Comment by blochl** [comment]
@ncw , thanks again for your answer.
Yes, currently I am interested only in Google Drive. Yes, displaying a URL and asking for the code is exactly what is missing for me.

Just out of curiosity though, I wonder why it won't work for other providers? Why redirect should interfere? After all, I'm looking for decoupling of the authentication from the rest of the configuration. Does this also mean that non-interactive setup doesn't work for other providers?

**Comment by ncw** [comment]
> Yes, currently I am interested only in Google Drive. Yes, displaying a URL and asking for the code is exactly what is missing for me.

With a very small patch I managed to make it work like this - I had to type in the verification code.

Note the `--auto-confirm` which takes the default for every interactive question.

```
$ rclone config --config /tmp/z.conf --auto-confirm create drive2 drive
Remote config
If your browser doesn't open automatically go to the following link: https://accounts.google.com/o/oauth2/auth?access_type=offline&client_id=YYY.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&state=XXX
Log in and authorize rclone for access
Enter verification code> **VERIFICATION CODE PRINTED IN THE BROWSER TYPED IN BY ME**
Paste the following into your remote machine --->YYY
{"access_token":"XXX","token_type":"Bearer","refresh_token":"YYY","expiry":"2018-10-15T16:52:09.169776254+01:00"}
<---End pasteConfigure this as a team drive?
--------------------
[drive2]
type = drive
config_automatic = yes
token = {"access_token":"XXX","token_type":"Bearer","refresh_token":"YYY","expiry":"2018-10-15T16:52:09.169776254+01:00"}
--------------------
```

> Just out of curiosity though, I wonder why it won't work for other providers? Why redirect should interfere? After all, I'm looking for decoupling of the authentication from the rest of the configuration. Does this also mean that non-interactive setup doesn't work for other providers?

The full oauth flow requires the users browser to redirect to the "Redirect URL`.  Google support a special redirect URL

```
	// TitleBarRedirectURL is the OAuth2 redirect URL to use when the authorization
	// code should be returned in the title bar of the browser, with the page text
	// prompting the user to copy the code and paste it in the application.
	TitleBarRedirectURL = "urn:ietf:wg:oauth:2.0:oob"
```

This displays a code in the browser (and in the title bar).  This is a google only extension to oauth, everyone else requires the RedirectURL and the webserver listening on that URL.

Here is the patch

```go
--- a/lib/oauthutil/oauthutil.go
+++ b/lib/oauthutil/oauthutil.go
@@ -361,7 +361,7 @@ func doConfig(id, name string, m configmap.Mapper, errorHandler func(*http.Reque
 			return PutToken(name, m, token, false)
 		}
 	case TitleBarRedirectURL:
-		useWebServer = automatic
+		useWebServer = false
 		if !automatic {
 			fmt.Printf("Use auto config?\n")
 			fmt.Printf(" * Say Y if not sure\n")
```

I think that needs to be controlled with a flag- does it look like it does the right thing?

BTW the google recommended way of doing this is using a service account which is fully supported with `rclone config create` though it could probably do with an option not to run the oauth stuff in that case.

**Comment by blochl** [comment]
Hi @ncw ,
Sorry for the delay - was busy with other things in the past few days.
The last solution doesn't do quite what I need: I don't want to use `--auto-confirm`, as I need some non-default options. I want to use a ready-made config file, and only do the authentication part separately.

However, I was able to do (almost) what I need using your previous suggestion:

```rclone config update remoteName token ''```

Notice, that I have left the token field empty. Now it asks me (as expected) "`Use auto config?`", I answer "`n`", go to the link, copy-paste the token, **perfect**!!!

Why "**almost** what I need", if it's perfect? Because then it asks me: 

```Change current team drive ID "XXXXXX"?```

... And I would like to avoid this question if possible, because it is not needed - I do not want to change the drive ID, and I did not ask for it in the command - I have only asked to change the token.

Is it possible to get rid of this extra question somehow?

But overall - problem solved! Thanks a lot for your advice!
